### ISSUE TITLE
Code that moves and cleans up objects as they move from /deposit to storage roots doesn't always work as expected

### ISSUE THREAD

**Issue by julianmorley** [sul-dlss/preservation_robots#307]
**Describe the bug**
The problem documented in https://github.com/sul-dlss/preservation_catalog/issues/1633 still occurs in production.

**User Impact**
Accessioning experiences issues (delays, failed accessioning, other errors), especially during periods of heavy activity.

**Expected behavior**
Files in /deposit should be consistently moved into pres storage roots without errors, and any artifacts left in /deposit should be consistently cleaned up.

**Additional context**
This issue is caused by how Ceph handles the creation and deletion of hard links; these actions are asynchronous and during periods of heavy activity may take seconds to complete. So if the code is trying to remove the object from /deposit, thinking that it has been successfully moved to the storage root, but the object has not yet finished creating the new links, a read/permissions occur will occur. Sometimes this is resolved by the passage of time, but a more consistent fix is to restart the metadata server managing that storage root.

It looks like this:
```
ls: cannot access /pres-01/sdr2objects/qt/116/cp/8268/qt116cp8268/v0004/manifests/versionAdditions.xml: Permission denied
total 22
-rw-rw-r-- 1 pres pres 6733 Jan 24 09:28 fileInventoryDifference.xml
-rw-rw-r-- 1 pres pres 1606 Jan 24 09:28 manifestInventory.xml
-rw-rw-r-- 1 pres pres 7971 Jan 24 09:28 signatureCatalog.xml
-????????? ? ?    ?       ?            ? versionAdditions.xml
-rw-rw-r-- 1 pres pres 4634 Jan 24 09:28 versionInventory.xml
```

**Comment by julianmorley** [comment]
I suspect that these zipmaker errors https://app.honeybadger.io/projects/54415/faults/69368144 are also related, since they suggest zip's inability to read data from Ceph storage immediately after the file is placed there by Robots (the file is there, and is readable, by the time I hand-check it, which implies a transitory, immediate failure to read that might have been solved by a larger delay between deposit and zipmaker).

**Comment by jmartin-sul** [comment]
> I suspect that these zipmaker errors https://app.honeybadger.io/projects/54415/faults/69368144 are also related, since they suggest zip's inability to read data from Ceph storage immediately after the file is placed there by Robots (the file is there, and is readable, by the time I hand-check it, which implies a transitory, immediate failure to read that might have been solved by a larger delay between deposit and zipmaker).

Yeah, I definitely share that suspicion, and I think our working hypothesis about the cause of these zipmaker errors has been something along these lines (i.e., something a bit handwavey about robot creation/linking of the moab files from the deposit space and the interplay of that with Ceph MDS)?

And while it's not something I've collected stats on, my recollection from looking at the pres cat failure queues regularly is that the majority of zipmaker errors are of this sort, and that most of the rest of the zipmaker errors are a similar sort of thing where the zip binary actually errors loudly during zip creation because of an inability to read the file (whereas the HB alert linked here indicates that the zip binary warned about inability to read a file, but the job failure was caused by our proactive size check after zip creation).

**Comment by jmartin-sul** [comment]
> The actual fix is probably in preservation_robots; see #277 for context. Some additional logic or checks or re-ordering of actions in pres robots would probably address this. Maybe something like
> 
>     * create hardlinks in destination storage root,
> 
>     * check that all are readable in the destination (and keep trying until they arre)
> 
>     * remove hardlinks from source, check that they're all removed
> 
>     * _then_ report that the object has been accessioned to preservation catalog for zipmaker etc to do work

so... i'm actually not sure how close we are to what the above describes without digging into the moab-versioning code a bit more.  it seems... plausible that we're already doing something like this from the pres robots side of things.  a bit more detail:

relevant workflow steps:
```xml
<workflow-def id="preservationIngestWF" repository="sdr">
  <process name="start-ingest" sequence="1" status="completed">
    <label>Initiate preservation ingest workflow</label>
  </process>
  <process name="transfer-object" sequence="2">
    <label>Get bagit/Moab deposit bag for new Moab version from (common-accessioning sdrIngestTransfer robot)</label>
    <prereq>start-ingest</prereq>
  </process>
  <process name="validate-bag" sequence="3">
    <label>Verify the bagit/Moab deposit bag structure of new version</label>
    <prereq>transfer-object</prereq>
  </process>
  <process name="update-moab" sequence="4">
    <label>Create/update Moab object from deposit bag</label>
    <prereq>validate-bag</prereq>
  </process>
  <process name="update-catalog" sequence="5">
    <label>Create/update Preservation Catalog entry</label>
    <prereq>update-moab</prereq>
  </process>
  <process name="complete-ingest" sequence="6">
    <label>Clean up workspace; transfer control back to accessioning</label>
    <prereq>update-catalog</prereq>
  </process>
</workflow-def>
```

so before the hardlink creation we "Transfer and untar the object from the DOR export area to the SDR deposit area.":  https://github.com/sul-dlss/preservation_robots/blob/1c1032af89f6fabe84831a419b66fac5c24b79ef/lib/robots/sdr_repo/preservation_ingest/transfer_object.rb#L31-L38

then we validate the bag:  https://github.com/sul-dlss/preservation_robots/blob/1c1032af89f6fabe84831a419b66fac5c24b79ef/lib/robots/sdr_repo/preservation_ingest/validate_bag.rb#L24-L29

>     * create hardlinks in destination storage root,

i believe that happens here ("Robot for ingesting deposit bag into Moab object (creating a new version)"):  https://github.com/sul-dlss/preservation_robots/blob/1c1032af89f6fabe84831a419b66fac5c24b79ef/lib/robots/sdr_repo/preservation_ingest/update_moab.rb#L22-L30

>     * check that all are readable in the destination (and keep trying until they arre)

this is the part i'm a bit unclear on atm... the above `UpdateMoab#update_moab` calls `Moab::StorageObjectVersion#verify_version_storage`.  and the comments for that moab-versioning method lead one to believe that the files on disk are being checked, but a cursory glance at that method and a couple levels into the helper methods it calls don't make it apparent that anything other than the data stream files are examined.

at present, we then tell pres cat the moab has been updated:  https://github.com/sul-dlss/preservation_robots/blob/1c1032af89f6fabe84831a419b66fac5c24b79ef/lib/robots/sdr_repo/preservation_ingest/update_catalog.rb#L35-L43

>     * remove hardlinks from source, check that they're all removed

at present that happens here:  https://github.com/sul-dlss/preservation_robots/blob/1c1032af89f6fabe84831a419b66fac5c24b79ef/lib/robots/sdr_repo/preservation_ingest/complete_ingest.rb#L23-L28

note that this step _definitely_ walks the new moab directory (albeit from the pres robots VM) _after_ the deposit bag is unlinked:  https://github.com/sul-dlss/preservation_robots/blob/1c1032af89f6fabe84831a419b66fac5c24b79ef/lib/robots/sdr_repo/preservation_ingest/complete_ingest.rb#L30-L46

to my knowledge, we haven't been seeing failures on the `complete-ingest` step, which would indicate to me that pres robots is consistently able to read the new moab directory it linked, even right after the original (in deposit bag) files were already unlinked.  more on this a in a bit...

>     * _then_ report that the object has been accessioned to preservation catalog for zipmaker etc to do work

and in the current way of doing things, this last proposed step already happened.


so, to recap, what we're doing now, as i understand it:
1. transfer (via tarpipe) the content from dor export dir to sdr deposit dir
2. validate the contents of the deposit bag
3. hard link the deposit bag contents to the new moab version directory (the pres robots `update_moab` method calling moab-versioning's `ingest_bag`
  1. ?? walk the newly created moab dir's contents?  not clear to me atm, depends on what moab-versioning's `verify_version_storage` is actually doing.  even if it is walking that moab dir, it's only doing it from the pres robots VM.
4. tell pres cat the new moab version is read (pres robots' `update_catalog`)
5. `complete-ingest`: pres robots unlinks the deposit bag and immediately makes sure the new moab version's files are readable (again, from the pres robots VM).

so, it would seem to me that the delta between what we're doing, and what @julianmorley is proposing is that we may not be walking the new moab version's contents at all before the `update_catalog` step (which is what triggers pres cat's replication pipeline, which reads moabs to create zips from the pres cat boxes, which fails sometimes as in the HB alert julian linked).

i would make two conjectures based on the above:
1. it's not clear to me that walking the files from the pres robots side before `update-catalog` will help, even if we aren't presently doing that in `update-moab` (via `verify_version_storage`).  my reasoning: if the files were unreadble from pres robots (as they are for pres cat's zipmaker) _after_ the deposit bag is removed/unlinked, we would expect to see the `complete-ingest` step fail with about the same frequency as we see zipmaker fail to read.  does that sound correct?  but to my knowledge, we rarely if ever see `complete-ingest` fail, correct?
2. the most valuable additional check we can add based on what we currently know is a walk of the new moab dir _from a pres cat VM_ before pres cat updates its database entry for the druid/moab.

and so i would refine julian's proposal to suggest that we "check that all are readable in the destination" from pres cat, before the database is updated.

i can imagine two easy-ish ways to do this:
1. just roll that directory walk into pres cat's `CatalogController#update` behavior (preferably via a private helper method or a service class and not directly in the `update` method).
2. create a new REST method on the pres cat side that does that directory walk and reports the result.  call that from pres robots `UpdateCatalog#update_catalog` method, where we currently call `wait_as_needed` (see also #277).
  * an alternate version of this would be to make this check a new workflow step, not part of the existing `update-catalog` step, run before the `update-catalog` step (but after `update-moab`).

that proposal would get us a walk of the moab directory from the pres cat VMs, before the catalog is updated, and before the bag is unlinked.

the nice things about doing that check on the pres cat side at the top of `update-catalog`, before the actual pres cat DB is updated, are that
* it would more closely do a dry run of the behavior that's erroring (i.e. reading the moab contents from a pres cat VM)
* it would block the catalog update that would trigger replication in a case where the read check fails
* it would be implemented in such a way that the robot step that failed could be mindlessly retried (doing it at the top of `update-catalog` would cause the error to happen before the non-idempotent part of the `update)

adding a new workflow step would, i think, be slightly clearer/cleaner than stuffing this in at the beginning of `update-catalog`, though i think it's also slightly more development and testing overhead to do it that way.  and putting this check at the top of `update-catalog` does not seem egregiously wrong to me, from a semantic perspective.

so, it seems to me that we can do something like what julian suggests, without a reshuffling of workflow steps, just the insertion of a directory walk that's not there.  which makes me a bit less worried about the regression testing version (though i think we'd still want to exercise this thoroughly in stage before rolling it out to prod).

all of the above also implies that there _is_ some code change in preservation_catalog, since i think that's where we want to do the pre-update-catalog moab dir walk, so maybe i shouldn't have transferred this issue to the pres robots repo afterall 😄 (we'll see, i sorta expect minor changes in both codebases for this).

corrections welcome, definitely some tricky/unclear interactions there, and i will proofread this momentarily...

**Comment by jmartin-sul** [comment]
i realize, after re-reading my last comment, that my proposal says "and before the bag is unlinked" for the moab walk from pres cat... which does not get at the crux of what @julianmorley is asking for, which is to only update the preservation catalog database _after both_ successful unlinking and moab walking. 🤦 

but also, the change i'm proposing feels like it has less regression risk and testing burden to me, and also is likely lower effort, and... _may_ still be worth trying?  maybe accessing the files from the pres cat VM before the unlinking will make those pres cat Ceph MDS instances aware of the files when they would not otherwise have been?  grasping at straws a bit there as i'm not so familiar with the ceph internals.

anyway, very interested in feedback from both @julianmorley and @andrewjbtw on the prior comment.

**Comment by julianmorley** [comment]
I've been doing my own attempts to re-create this file-read issue in our test Chaos environment, but so far with no success. I'm slightly leaning towards this now being a concurrency issue (Robots not giving up a write cap  quickly enough when a pres VM asks for a read cap on the same file), since most the HB errors I can find that smell of this seem to be happening from a Pres VM immediately after Robots is done with accessioning. I have some more dirty tests I want to try to see if I can invoke this behavior.

FWIW I've been able to completely blitz our test cluster with 300 files/sec link/unlink without seeing any issues, and these boxes have less resources and slower disk than production, so I'm no longer certain that it's some kind of back-office asynch work with the object metadata that occasionally causes this.  But the big proviso there is that we're now running a different version of Ceph and all forms of these errors have dropped off since the upgrade at the end of March, so it's possible some patch btw Ceph `15.2.5` and `15.2.9` has also helped.

**Comment by julianmorley** [comment]
As an aside, I'm really grateful for the code walkthru above, @jmartin-sul! I tried spelunking the code on Friday and found some of this, but your recap is much more comprehensive and fills in some crucial gaps.

**Comment by jmartin-sul** [comment]
> But the big proviso there is that we're now running a different version of Ceph and all forms of these errors have dropped off since the upgrade at the end of March, so it's possible some patch btw Ceph `15.2.5` and `15.2.9` has also helped.

nice, my gut feeling from watching the resque dashboard intermittently was that we're seeing a lot less of this lately, which i'd been meaning to ask about, so good to know that wasn't just me imagining things.

**Comment by jmartin-sul** [comment]
> As an aside, I'm really grateful for the code walkthru above, @jmartin-sul! I tried spelunking the code on Friday and found some of this, but your recap is much more comprehensive and fills in some crucial gaps.

oh nice, very glad that was useful!  hadn't looked at any of this since last fall, so needed a little refresher myself.

**Comment by jmartin-sul** [comment]
is there anything flush-like or commit-like that can be done so that pres robots can signal "all done with the files in the moab directory" after it writes those?

**Comment by julianmorley** [comment]
> is there anything flush-like or commit-like that can be done so that pres robots can signal "all done with the files in the moab directory" after it writes those?

Not without doing some funky librados coding, I think. This may all boil down to a load condition where Robots has a huge bucket of caps, pres VMs want access to some of them, and it takes just a little bit too long for them to sort out who can read/write what that some of the pres VM's file requests get denied. I suspect that if we increased the pause from it's current 0.05 sec value we'd probably see less errors, as that'd give the clients more time to negotiate access.

**Comment by andrewjbtw** [comment]
Thanks for all the analysis and the code walkthrough above. I've never really followed the zipmaker side of this in much detail. In accessioning, the sequence of events I've seen has always been the following:

1. User makes a new version of something that hasn't been updated in a while (for values of "while = >1 day")
2. User realizes they have more changes to make, waits for accessioning to finish so they can version again.
3. User creates next new version, within minutes of previous version finishing
4. Permission denied error at one of the following steps
 - the content diff at the shelve[?] step when there's a prescat call to check on the contents of the preserved object and compare with what's in the dor workspace to be accessioned https://app.honeybadger.io/projects/54415/faults/65631904
 - the update-catalog permissioned denied errors we seem to see more often

The most recent occurrence of this issue was when Jessica updated metadata on a batch of objects, realized she needed to make another update, and then ran into errors with multiple items the second time around. Previous cases have similar usage patterns when I've looked into it with the user.

I don't think I've ever seen this error when a user has created only version N. It's always when a version (N + 1) has been created a short period of time after creating version N. And the permission denied error seems to be when reading a file from version N, not a file from version (N + 1).

My guess from John's outline above is that the update-moab step, like whichever step runs the content diff, is hitting permission denied while reading a file from the previous version, not from any of the files in the current version. Maybe update-moab walks the previous version in order to determine what the new version directory should contain. Could it be verifying deduplication took place correctly when the transferred bag was created and that only new and/or modified files are being stored in the new version directory? Building the new manifest files from the old ones?

People making a new version, then making another one right afterwards is not a high frequency event, which makes it hard to gather statistics on how often there's an accessioning error due to permission denied. Measured against all accessioning, we don't see the issue that often. But measured against people making multiple versions in a short period of time, it happens more often. But when it does happen, it blocks further edits to that object until it's fixed. The zip errors seem more common, presumably because there's always a near-term follow-up read to create the zip, while there isn't always a follow up to make a new object version.

**Comment by julianmorley** [comment]
The other thing I just thought about looking at this sequencing:

```
  <process name="update-moab" sequence="4">
    <label>Create/update Moab object from deposit bag</label>
    <prereq>validate-bag</prereq>
  </process>
  <process name="update-catalog" sequence="5">
    <label>Create/update Preservation Catalog entry</label>
    <prereq>update-moab</prereq>
  </process>
  <process name="complete-ingest" sequence="6">
    <label>Clean up workspace; transfer control back to accessioning</label>
    <prereq>update-catalog</prereq>
  </process>
```

If update-catalog is async and returns quickly (before zipmaker has done its thing) it might be possible that the unlinking of the original content in `/deposit` (and Ceph's subsequent need to create a new object on the back end to track the change) might be causing file permission problems - a file gets unlinked whilst zipmaker is trying to work on it. I believe that was my original reasoning for wanting to change the order of the workflow.

**Comment by jmartin-sul** [comment]
So... `update-catalog` is synchronous on the pres robots side relative to what the REST-invoked pres cat method does:  `Preservation::Client.update` (called in the robot step linked above) calls [`CatalogController#update`](https://github.com/sul-dlss/preservation_catalog/blob/2dc297dd416f5264902fabb919805cac666a26db/app/controllers/catalog_controller.rb#L29-L45), which does a cursory scan of the moab structure (including version check) and updates the database.  It then returns the HTTP status code corresponding to the outcome of trying to do those things.

But, if that's successful, the `ZipmakerJob` is queued after the update is saved to the (pres cat) database.  At the same time, `update-catalog` having succeeded will allow `complete-ingest` to start.  So there is definitely a race condition between the two async operations that are directly gated on a successful `update-catalog` step:  pres cat workers reading to make archive zips and pres robots cleaning up the deposit directory.

This is the post-save hook in pres cat that starts the replication pipeline for the druid version after the catalog has been updated:  https://github.com/sul-dlss/preservation_catalog/blob/2dc297dd416f5264902fabb919805cac666a26db/app/models/preserved_object.rb#L14

> The delay timer we shimmed into `update-catalog` was a dirty way of implementing option A, but that's obviously imperfect and will both cover druids that don't need it and sometimes miss druids that do, which was why I was advocating for putting `update-catalog` *after* `complete-ingest`, or performing the unlinking of `/deposit` content prior to calling `update-catalog`.

I don't disagree.  I'd just worry about regression risk here, and would want to be mindful of the need for adequate testing in stage and observation in prod if we made that change.

**Comment by jmartin-sul** [comment]
FWIW, looking at that robot code again, I actually can't think of misbehavior that'd be caused by moving the deposit content unlinking to the top of the `update-catalog` step, but I haven't thought about it a ton.

**Comment by jmartin-sul** [comment]
met with @andrewjbtw, @julianmorley, and @vivnwong this morning, and we decided to move forward as follows:

1. We'll move the call to unlink the deposit directory to the beginning of the pres robots `update-catalog` step.  None of us can think of any problems this would introduce, and while stage is not similar enough to prod in terms of Ceph setup to give us an idea of whether this change will fix the problem, we are reasonably confident that stage testing can at least give some assurance that we're not introducing regressions.  Our main worries, in terms of regression risk, are oversights and unknown-unknowns in an important part of our repository infrastructure.
2. We'll get the week's first-responder to shepherd the work.  If need be, they can pair w/ @jmartin-sul on the code change.  Whoever does the implementing should probably also meet w/ @andrewjbtw for 30 to 60 min to write up a thorough manual test plan (which I suspect will be a combination of the manual testing that Andrew describes above, as well as a run of the integration test suite, and a manual look at the preserved and replicated content created by infra integration tests, since the tests don't actually check preserved/replicated content created for the test SDR objects).  We'll also let it run in stage for a a few days, which should allow gbooks and other testing to more naturally exercise the change.
3. I'll spin off a separate pres robots ticket for the FR work, with a more straight-forward TODO checklist, and a link back to this ticket for more context.
4. I'll create a separate https://github.com/sul-dlss/infrastructure-integration-test/ for improving the test suite to have some basic expectations about preserved content, so that the infra integration tests can more easily be used to check pres changes with less manual follow up.

**Comment by jmartin-sul** [comment]
more focused ticket for the code change:  https://github.com/sul-dlss/preservation_robots/issues/310

infra integration test ticket:  https://github.com/sul-dlss/infrastructure-integration-test/issues/248

**Comment by jmartin-sul** [comment]
One last thing for now:  I'd meant to reply to @andrewjbtw's last comment about the zipmaker errors, vs the content diff and `update-catalog` errors:  I'd been more focused on the zipmaker issues, as I'd been seeing them at much higher rates.  Tens to hundreds per week at worst, vs (to my recollection) a few every couple weeks for `update-catalog` and content diff Ceph read errors.

We are somewhat hopeful that the fix proposed in #310 will ameliorate both sorts of Ceph read issue, but we'll see.  Personally, based on what @julianmorley has explained, I'm slightly more confident that #310 will address the zipmaker issues than I am that it'll address the content diff and `update-catalog` read issues, but we don't expect #310 to cause any harm on that front, at least.

**Comment by ndushay** [comment]
FYI:  I don't believe the actual fix was applied until 2 weeks ago. See https://github.com/sul-dlss/preservation_robots/pull/311 merge date - I think John was working on the PR for a while.

**Comment by justinlittman** [comment]
I believe the work on this ticket is completed; if you disagree, feel free to re-open and describe any additional work.
